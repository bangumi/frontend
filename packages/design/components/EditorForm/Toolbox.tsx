import type { FC } from 'react';
import React, { memo } from 'react';

import { Bold, Image, Italic, Link, Size, Underscore } from '@bangumi/icons';

interface ToolboxProps {
  handleClickEvent: (type: string, payload?: unknown) => void;
  style?: React.CSSProperties;
}

const Toolbox: FC<ToolboxProps> = ({ handleClickEvent, style }) => {
  return (
    <div className='bgm-editor__toolbox' style={style}>
      <Bold
        data-testid='bold'
        onClick={() => {
          handleClickEvent('bold');
        }}
        title='粗体 [Ctrl+B]'
      />
      <Italic
        data-testid='italic'
        onClick={() => {
          handleClickEvent('italic');
        }}
        title='斜体 [Ctrl+I]'
      />
      <Underscore
        data-testid='underscore'
        onClick={() => {
          handleClickEvent('underscore');
        }}
        title='下划线 [Ctrl+U]'
      />
      <Image
        data-testid='image'
        onClick={() => {
          handleClickEvent('image');
        }}
        title='图片 [Ctrl+P]'
      />
      <Link
        data-testid='link'
        onClick={() => {
          handleClickEvent('link');
        }}
        title='链接 [Ctrl+L]'
      />
      <Size
        data-testid='size'
        onClick={() => {
          handleClickEvent('size');
        }}
        title='文字尺寸 [Ctrl+S]'
      />
    </div>
  );
};

export default memo(Toolbox);
