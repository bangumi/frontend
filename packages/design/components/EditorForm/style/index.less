@import '../../../theme/base';
@import (reference) '../../../components/Form/style/index';

@toolbox-width: 215px;

.bgm-editor {
  &__container {
    display: inline-flex;
    flex-direction: column;
    min-height: 162px;
    border: 2px solid @gray-10;
    border-radius: 12px;
    padding: 8px 6px 6px;
    box-sizing: border-box;
    width: 100%;
  }

  &__toolbar {
    display: flex;
    justify-content: space-between;
    margin-bottom: 10px;
  }

  &__toolbox {
    display: flex;
    justify-content: space-between;
    align-items: center;
    height: 26px;
    width: @toolbox-width;
    padding: 0 8px;

    > svg {
      cursor: pointer;
    }
  }

  &__wordcount {
    display: flex;
    align-items: center;
    font-size: 14px;
    color: @gray-60;
    margin-right: 9px;
    white-space: nowrap;
  }

  &__text {
    flex: 1 0 auto;
    padding: 0 6px;
    resize: vertical;
    border: none;
    outline: none;
    color: @gray-100;
    line-height: 26px;
    font-size: 16px;

    &::placeholder {
      color: @gray-60;
    }
  }

  &__form {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    justify-content: flex-start;
  }

  &__submit {
    font-size: 14px;
    display: flex;
    margin-top: 12px;
    align-items: center;
    justify-content: flex-start;
    gap: 10px;

    .bgm-editor__bbcode-tip {
      color: @gray-60;
      line-height: 24px;

      &__left {
        .sm-screen-hidden();
      }
    }
  }

  &__button {
    padding: 0 15px;
  }

  .bgm-form &__text {
    width: auto;
  }

  // 在 Form 中使用 Editor 而非 EditorForm 的场景
  .bgm-form--compact > &__container {
    .compact-form-item();
  }

  .bgm-form--compact > &__form {
    align-items: stretch;
    width: auto;

    .bgm-editor__container {
      .no-top-border();
    }
  }
}
