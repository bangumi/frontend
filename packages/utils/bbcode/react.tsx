import React from 'react';

import { convert } from './convert';
import { Parser } from './parser';
import type { CodeNodeTypes, NodeTypes, VNode } from './types';

/* eslint-disable no-useless-escape */
const toCamelCase = (s: string) => s.replace(/(\-\w)/g, (k) => k[1]?.toUpperCase() ?? '');

const joinClassName = (cls: string | string[]) => {
  if (Array.isArray(cls)) {
    return cls.join(' ');
  }
  return cls;
};

const convertToReactProps = (node: VNode, key?: React.Key) => {
  const { props, style, className } = node;

  const transformedStyle: Record<string, string> = style
    ? Object.entries(style).reduce((pre, [key, val]) => {
        return {
          ...pre,
          [toCamelCase(key)]: val,
        };
      }, {})
    : {};

  const transformedClassName = className !== undefined ? joinClassName(className) : undefined;

  const reactProps = {
    ...props,
    style: transformedStyle,
    className: transformedClassName,
    key,
  };

  return reactProps;
};

export const renderNode = (node: NodeTypes, key?: React.Key): React.ReactElement | string => {
  if (typeof node === 'string') {
    const splittedStr: Array<string | React.ReactElement> = node.split(/\n/g);

    if (splittedStr.length < 2) {
      return splittedStr[0] ?? '';
    }
    const children = splittedStr.reduce<Array<string | React.ReactElement>>((pre, curr, idx) => {
      if (idx > 0) {
        return [...pre, React.createElement('br', null), curr];
      }
      return [...pre, curr];
    }, []);
    return React.createElement(React.Fragment, { key }, ...children);
  }

  const { type, children } = node;

  if (children) {
    return React.createElement(
      type,
      convertToReactProps(node, key),
      ...children.map((child, idx) => renderNode(child, idx)),
    );
  }
  return React.createElement(type, convertToReactProps(node, key));
};

export const render = (rawStr: string): Array<React.ReactElement | string> => {
  const nodes: CodeNodeTypes[] = new Parser(rawStr).parse();

  return nodes.map((node, idx) => renderNode(convert(node, {}), idx));
};
