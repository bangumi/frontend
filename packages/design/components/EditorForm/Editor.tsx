import { size } from 'lodash';
import type { FocusEventHandler } from 'react';
import React, {
  forwardRef,
  useCallback,
  useEffect,
  useImperativeHandle,
  useRef,
  useState,
} from 'react';

import Toolbox from './Toolbox';

export interface EditorProps {
  placeholder?: string;
  /** 是否显示工具栏 */
  showToolbox?: boolean;
  /** 是否显示字数统计 */
  showWordCount?: boolean;
  /** textarea 通过键盘按下提交触发事件 */
  onConfirm?: (value: string) => void;
  /**
   * @default empty string
   */
  value?: string;
  /** 内容改变时的回调函数 */
  onChange?: (value: string) => void;
  /** 在被渲染时是否自动获取焦点 */
  autoFocus?: boolean;
  /** 文本域初始行数 */
  rows?: number;
  name?: string;
  onBlur?: FocusEventHandler<HTMLTextAreaElement>;
}

const keyToEvent: Record<string, string> = {
  b: 'bold',
  i: 'italic',
  u: 'underscore',
  l: 'link',
  s: 'size',
  p: 'image',
} as const;

const Editor = forwardRef<HTMLTextAreaElement, EditorProps>(
  (
    {
      placeholder,
      showToolbox = true,
      showWordCount = true,
      onConfirm,
      value = '',
      onChange,
      autoFocus,
      rows,
      name,
      onBlur,
    },
    ref,
  ) => {
    const [selection, setSelection] = useState<[number, number]>();
    const innerRef = useRef<HTMLTextAreaElement>(null);

    useImperativeHandle<HTMLTextAreaElement | null, HTMLTextAreaElement | null>(
      ref,
      () => innerRef.current,
    );

    useEffect(() => {
      // 用来确保输入光标在自动插入的quote内容之后。
      const el = innerRef.current;
      el?.setSelectionRange(el.value.length, el.value.length);
    }, []);

    // content改变时应重新设置选择范围
    useEffect(() => {
      if (!selection) return;
      const [start, end] = selection;
      const el = innerRef.current;
      el?.setSelectionRange(start, end);
      el?.focus();
    }, [selection]);

    const updateContent = useCallback(
      (newContent: string): void => {
        setSelection(undefined);
        onChange?.(newContent);
      },
      [setSelection, onChange],
    );

    const insertTag = useCallback(
      (el: HTMLTextAreaElement, prefix: string, suffix: string, selected = ''): void => {
        const preStart = el.selectionStart;
        const preEnd = el.selectionEnd;
        const preValue = el.value;
        updateContent(
          `${preValue.slice(0, preStart)}${prefix}${selected}${suffix}${preValue.slice(preEnd)}`,
        );
        // 未选择内容，插入BBCode标签后应将光标置于标签中间
        if (selected.length === 0) {
          const cursorPos = preStart + prefix.length + selected.length;
          setSelection([cursorPos, cursorPos]);
        }
        el.focus();
      },
      [updateContent, setSelection],
    );

    // 使用useCallback避免在内容改变时触发Toolbox的渲染
    const handleToolboxEvent = useCallback(
      (type: string, payload?: unknown): void => {
        const el = innerRef.current;
        if (!el) {
          return;
        }
        const preStart = el.selectionStart;
        const preEnd = el.selectionEnd;
        const selected = el.value.slice(preStart, preEnd);
        switch (type) {
          case 'bold': {
            insertTag(el, '[b]', '[/b]', selected);
            break;
          }
          case 'italic': {
            insertTag(el, '[i]', '[/i]', selected);
            break;
          }
          case 'underscore': {
            insertTag(el, '[u]', '[/u]', selected);
            break;
          }
          case 'image': {
            const value = prompt('请输入图片链接');
            if (value === null) break;
            insertTag(el, '[img]', '[/img]', value);
            break;
          }
          case 'link': {
            const value = prompt('请输入链接地址');
            if (value === null) break;
            const prefix = `[url=${value}]`;
            const suffix = '[/url]';
            insertTag(el, prefix, suffix, selected || '链接描述');
            if (!selected) {
              // 选中“链接描述”
              const start = preEnd + prefix.length;
              const end = start + 4;
              setSelection([start, end]);
            }
            break;
          }
          case 'size': {
            const value = prompt('请输入字体大小');
            if (value === null) break;
            insertTag(el, `[size=${value}]`, '[/size]', selected);
            break;
          }
        }
      },
      [innerRef, insertTag, setSelection],
    );

    const handleKeyDown = useCallback(
      (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
        // Ctrl + Enter & Alt + S 触发提交
        if (
          ((e.ctrlKey || e.metaKey) && e.key === 'Enter') ||
          (e.altKey && e.key.toLowerCase() === 's')
        ) {
          onConfirm?.(innerRef.current!.value);
          e.preventDefault();
          return;
        }
        // https://bgm.tv/help/bbcode
        if (e.ctrlKey || e.metaKey) {
          const key = e.key.toLowerCase();
          const event = keyToEvent[key];
          if (event) {
            handleToolboxEvent(event);
            e.preventDefault();
          }
        }
      },
      [onConfirm, handleToolboxEvent],
    );

    return (
      <div className='bgm-editor__container'>
        <div className='bgm-editor__toolbar'>
          {showToolbox && <Toolbox handleClickEvent={handleToolboxEvent} />}
          {showWordCount && value && (
            <div className='bgm-editor__wordcount'>已输入 {size(value)} 字</div>
          )}
        </div>
        <textarea
          className='bgm-editor__text'
          placeholder={placeholder}
          ref={innerRef}
          value={value}
          autoFocus={autoFocus}
          onChange={(e) => {
            updateContent(e.target.value);
          }}
          onKeyDown={handleKeyDown}
          rows={rows}
          name={name}
          onBlur={onBlur}
        />
      </div>
    );
  },
);

export default Editor;
