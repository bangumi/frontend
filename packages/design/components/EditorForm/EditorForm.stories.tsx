import type { Meta, StoryFn } from '@storybook/react';
import type { FC } from 'react';
import React, { useState } from 'react';

import EditorForm from '.';
import Editor from './Editor';
import Toolbox from './Toolbox';

const componentMeta: Meta<typeof EditorForm> = {
  title: 'Modern/EditorForm',
  component: EditorForm,
  subcomponents: {
    Toolbox,
    Editor,
  } as Record<string, FC<unknown>>,
};

export default componentMeta;

const Template: StoryFn<typeof EditorForm> = (args) => {
  const [value, setValue] = useState(args.value);
  return <EditorForm {...args} value={value} onChange={setValue} />;
};

export const Usage = Template.bind({});

Usage.args = {
  placeholder: '请输入内容',
};
