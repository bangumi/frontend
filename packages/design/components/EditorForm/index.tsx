import './style';

import classnames from 'classnames';
import React, { forwardRef } from 'react';

import Button from '../Button';
import Link from '../Typography/Link';
import type { EditorProps } from './Editor';
import Editor from './Editor';

export interface EditorFormProps extends EditorProps {
  /** 最外层 className */
  className?: string;
  /** 最外层 style */
  style?: React.CSSProperties;
  /** 确认按钮的文本 */
  confirmText?: string;
  /** 确认按钮后的回调 */
  onConfirm?: (content: string) => void;
  /** 取消按钮的文本 */
  cancelText?: string;
  /** 取消按钮的回调 */
  onCancel?: () => void;
  /**
   * 是否隐藏取消按钮
   * @default false
   */
  hideCancel?: boolean;
}

const EditorForm = forwardRef<HTMLTextAreaElement, EditorFormProps>(
  (
    {
      className,
      style,
      confirmText = '写好了',
      onConfirm,
      cancelText = '取消',
      onCancel,
      hideCancel = false,
      ...props
    },
    ref,
  ) => {
    return (
      <div className={classnames('bgm-editor__form', className)} style={style}>
        <Editor ref={ref} onConfirm={onConfirm} {...props} />
        <div className='bgm-editor__submit'>
          <Button
            color='blue'
            className='bgm-editor__button bgm-editor__button--confirm'
            onClick={() => onConfirm?.(props.value ?? '')}
          >
            {confirmText}
          </Button>
          {!hideCancel && (
            <Button type='text' className='bgm-editor__button' onClick={onCancel}>
              {cancelText}
            </Button>
          )}
          <span className='bgm-editor__bbcode-tip'>
            <span className='bgm-editor__bbcode-tip__left'>
              使用 Ctrl+Enter 或 Alt+S 快速提交 |{' '}
            </span>
            <Link isExternal to='https://bgm.tv/help/bbcode'>
              BBCode指南
            </Link>
          </span>
        </div>
      </div>
    );
  },
);

export default EditorForm;
