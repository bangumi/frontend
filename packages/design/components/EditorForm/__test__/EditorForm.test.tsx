import { fireEvent, render } from '@testing-library/react';
import React, { useState } from 'react';

import type { EditorFormProps } from '..';
import EditorForm from '..';

const TestEditorForm = (props: EditorFormProps) => {
  const [value, setValue] = useState('');
  return <EditorForm value={value} onChange={setValue} {...props} />;
};

describe('<EditorForm />', () => {
  it('render correctly with props', () => {
    const { asFragment } = render(
      <TestEditorForm className='custom class' placeholder='placeholder' confirmText='Confirm' />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('onConfirm event', () => {
    const onConfirm = vi.fn();
    const { getByText, getByPlaceholderText } = render(
      <TestEditorForm onConfirm={onConfirm} confirmText='Confirm' placeholder='placeholder' />,
    );
    const textarea = getByPlaceholderText('placeholder') as HTMLTextAreaElement;
    fireEvent.change(textarea, { target: { value: 'test' } });
    getByText('Confirm').click();
    expect(onConfirm).lastCalledWith('test');
  });

  it('onCancel event', () => {
    const onCancel = vi.fn();
    const { getByText } = render(<TestEditorForm onCancel={onCancel} />);
    getByText('取消').click();
    expect(onCancel).toHaveBeenCalled();
  });

  it('Ctrl + Enter & Alt + S should trigger onConfirm event', () => {
    const onConfirm = vi.fn();
    const { getByPlaceholderText } = render(
      <TestEditorForm onConfirm={onConfirm} placeholder='placeholder' />,
    );
    const textarea = getByPlaceholderText('placeholder') as HTMLTextAreaElement;

    fireEvent.change(textarea, { target: { value: 'test' } });
    fireEvent.keyDown(textarea, { key: 'Enter', ctrlKey: true });
    expect(onConfirm).lastCalledWith('test');

    fireEvent.change(textarea, { target: { value: 'test2' } });
    fireEvent.keyDown(textarea, { key: 's', altKey: true });
    expect(onConfirm).lastCalledWith('test2');
  });
});
