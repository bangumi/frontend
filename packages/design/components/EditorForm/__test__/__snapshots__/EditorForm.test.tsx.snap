// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<EditorForm /> > render correctly with props 1`] = `
<DocumentFragment>
  <div
    class="bgm-editor__form custom class"
  >
    <div
      class="bgm-editor__container"
    >
      <div
        class="bgm-editor__toolbar"
      >
        <div
          class="bgm-editor__toolbox"
        >
          <div
            data-testid="bold"
            title="粗体 [Ctrl+B]"
          />
          <div
            data-testid="italic"
            title="斜体 [Ctrl+I]"
          />
          <div
            data-testid="underscore"
            title="下划线 [Ctrl+U]"
          />
          <div
            data-testid="image"
            title="图片 [Ctrl+P]"
          />
          <div
            data-testid="link"
            title="链接 [Ctrl+L]"
          />
          <div
            data-testid="size"
            title="文字尺寸 [Ctrl+S]"
          />
        </div>
      </div>
      <textarea
        class="bgm-editor__text"
        placeholder="placeholder"
      />
    </div>
    <div
      class="bgm-editor__submit"
    >
      <button
        class="bgm-button bgm-editor__button bgm-editor__button--confirm bgm-button--primary bgm-button--color-blue"
      >
        Confirm
      </button>
      <button
        class="bgm-button bgm-editor__button bgm-button--text"
      >
        取消
      </button>
      <span
        class="bgm-editor__bbcode-tip"
      >
        <span
          class="bgm-editor__bbcode-tip__left"
        >
          使用 Ctrl+Enter 或 Alt+S 快速提交 | 
        </span>
        <a
          class="bgm-link"
          href="https://bgm.tv/help/bbcode"
        >
          BBCode指南
        </a>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
